gk/fib.c

```diff
@@ -1664,3 +1665,303 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	entry->stale = eth_cache->stale;
+	rte_memcpy(&entry->nexthop_ip, &eth_cache->ip_addr,
+		sizeof(entry->nexthop_ip));
+	rte_memcpy(&entry->l2_hdr, &eth_cache->l2_hdr, sizeof(entry->l2_hdr));
+	entry->ref_cnt = rte_atomic32_read(&eth_cache->ref_cnt);
+}
+
+static int
+fillup_grantor_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	rte_memcpy(&dentry->grantor_ip, &fib->u.grantor.gt_addr,
+		sizeof(dentry->grantor_ip));
+
+	fillup_gk_fib_ether_entry(fib->u.grantor.eth_cache, eth_tbl);
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gateway_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	fillup_gk_fib_ether_entry(fib->u.gateway.eth_cache, eth_tbl);
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_neighbor_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int num_entries = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	struct neighbor_hash_table *neigh_ht;
+	struct gk_fib_ether_entry *eth_tbl;
+	struct gk_fib_ether_entry *entry;
+
+	if (dentry->addr.proto == ETHER_TYPE_IPv4)
+		neigh_ht = &fib->u.neigh;
+	else
+		neigh_ht = &fib->u.neigh6;
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		fillup_gk_fib_ether_entry(eth_cache, entry);
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+int
+print_fib_dump_entry(void *fib_dump_entry, __attribute__((unused)) void *arg)
```

Comment:
`print_fib_dump_entry()` should be implemented in Lua, so `gkctl` can (1) filter the entries and the fields that must be printed and (2) have full control on the format of the output. Thus, `gkctl` can evolve without requiring patching Gatekeeper.

Reply:
Good point! However, I still need to find a way to implement this. Here is the problem: assuming we have a Lua function defined as ``function print_fib_dump_entry(...)`` in ``dylib.lua``. In gatekeeper dynamic configuration, we will create a Lua state, which will run ``example_of_dynamic_config_request.lua`` sent by the client. This lua script will call the C function ``list_fib_entries()``, which further calls the lua function ``function print_fib_dump_entry(...)``. That is, the C function will call a Lua function, so we need to obtain the Lua state which is [created dynamically](https://github.com/mengxiang0811/gatekeeper/blob/master/config/dynamic.c#L366) and holds the Lua function. Then, will we have to maintain the Lua state in the ``struct dynamic_config``, which seems unexpected?

Reply:
First of all, the Lua function `print_fib_dump_entry()` should be in `example_of_dynamic_config_request.lua` not in `dylib.lua` since `gkctl` should be able to have any print function it needs.

The problem you're pointing out is solved if you make `list_fib_entries()` a Lua function since it'll receive the Lua state and can easily call the Lua callback function.

Reply:
Yes, this is doable. However, this may requires to expose a lot of data structures to the Lua scripts, e.g., LPM table, spinlocks, gk fib, rte_lpm6_rule, lpm states, etc.

Reply:
I may be missing something here, but I see that only `struct gk_fib_ether_dump_entry` and `struct gk_fib_dump_entry` need to be exposed to the Lua script beside the interface of `list_fib_entries()` and the callback function.

Reply:
Maybe I miss something here, I thought to implement ``list_fib_entries()`` completely in Lua. If so, that's the case I described above. Do you mean that we can implement ``list_fib_entries(lua_State *l)`` as a C function, and then call it in Lua?

Reply:
Yes, implementing `list_fib_entries(lua_State *l)` as a C function, and then call it in Lua. We already have a number of such functions in Gatekeeper.